hack/generator/pkg/astmodel/storage_conversion_function.go

```diff
@@ -0,0 +1,345 @@
+// instance when we don't directly understand the structure of the parameter value.
+// To accommodate this, we first populate an intermediate form that is then converted.
+//
+// var staging IntermediateType
+// [copy values to staging]
+// staging.ConvertTo(parameter)
+//
+func (fn *StorageConversionFunction) generateIndirectConversionTo(receiver string, parameter string, ctx *CodeGenerationContext) []dst.Stmt {
+
+	local := fn.knownLocals.createLocal(receiver + "Temp")
+
+	parameterPackage := ctx.MustGetImportedPackageName(fn.staging.name.PackageReference)
+	localDeclaration := astbuilder.LocalVariableDeclaration(
+		local,
+		&dst.SelectorExpr{
+			X:   dst.NewIdent(parameterPackage),
+			Sel: dst.NewIdent(fn.staging.name.name),
+		},
+		fmt.Sprintf("// %s is our intermediate for conversion", local))
+	localDeclaration.Decorations().Before = dst.NewLine
+
+	callConvertTo := astbuilder.InvokeQualifiedFunc(
+		local, fn.name, dst.NewIdent(parameter))
+	callConvertTo.Decorations().Before = dst.EmptyLine
+	callConvertTo.Decorations().Start.Append(
+		fmt.Sprintf("// Populate %s from %s", parameter, local))
+
+	assignments := fn.generateAssignments(
+		dst.NewIdent(receiver),
+		dst.NewIdent(local),
+		ctx)
+
+	var result []dst.Stmt
+	result = append(result, localDeclaration)
+	result = append(result, assignments...)
+	result = append(result, callConvertTo)
+	return result
+}
+
+func (fn *StorageConversionFunction) generateAssignments(source dst.Expr, destination dst.Expr, ctx *CodeGenerationContext) []dst.Stmt {
+	var result []dst.Stmt
+
+	// Find all the properties for which we have a conversion
+	var properties []string
+	for p := range fn.conversions {
+		properties = append(properties, p)
+	}
+
+	// Sort the properties into alphabetical order to ensure deterministic generation
+	sort.Slice(properties, func(i, j int) bool {
+		return properties[i] < properties[j]
+	})
+
+	// Accumulate all the statements required for conversions, in alphabetical order
+	for _, prop := range properties {
+		conversion := fn.conversions[prop]
+		block := conversion(source, destination, ctx)
+		if len(block) > 0 {
+			//TODO: Tidy
+			firstStatement := block[0]
+			firstStatement.Decorations().Before = dst.EmptyLine
+			firstStatement.Decorations().Start.Append("// " + prop)
+			result = append(result, block...)
+		}
+	}
+
+	return result
+}
+
+func (fn *StorageConversionFunction) receiverName(receiver TypeName) string {
+	return fn.idFactory.CreateIdentifier(receiver.Name(), NotExported)
+}
+
+func (fn *StorageConversionFunction) parameterName() string {
+	if fn.conversionDirection == ConvertTo {
+		return "destination"
+	}
+
+	if fn.conversionDirection == ConvertFrom {
+		return "source"
+	}
+
+	panic(fmt.Sprintf("Unexpected conversion conversionType %v", fn.conversionDirection))
+}
+
+func (fn *StorageConversionFunction) createConversions(receiver TypeDefinition) []error {
```

Comment:
It might be worth chatting about this in person so I understand more fully how the `StoragePropertyConversion`s are used. I think they're set up the way they are because at construction time you don't have/know the final reciever name, so you need to have them be lambdas so that you can pick/choose what the receiver name actually is? The actual conversion though (say, optional int to int) is encoded at function construction time in the `StoragePropertyConversion` is that right?

Reply:
There **is** an assumption that the list of properties on both the receiver and other types won't change from the point where the analysis is done. 

I did it this way partly to keep the amount of work done in the code-generation phase as low as possible, but you do raise a possible risk.

Do you think we need to enforce this by doing a cross check at the codegen stage?